File: Python/LC/98.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ValidateBST(object):
    def isValidBST(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        result = []
        self.isValidateBST_brute(root, result)

        for i in range(1, len(result)):
            if result[i-1] > result[i]:
                return False

        return True

    def isValidateBST_brute(self, root, result):
        if root:
            self.isValidateBST_brute(root.left, result)
            result.append(root.val)
            self.isValidateBST_brute(root.right, result)

File: LC/98/98.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isValidBST(root: TreeNode | null): boolean {
    if (!root) return false;
    const traversals = [];
    inorderTraversal(root, traversals);
    return traversals.every((node, i) => i === 0 || traversals[i-1] < node);
};

function inorderTraversal(node: TreeNode, traversals: number[]): void {
    if (!node) return;
    inorderTraversal(node.left, traversals);
    traversals.push(node.val);
    inorderTraversal(node.right, traversals);
};
